File: services/payment/stripe_service.ts
import Stripe from 'stripe';

export class StripePaymentService {
  private stripe: Stripe;

  constructor(apiKey: string) {
    this.stripe = new Stripe(apiKey, { apiVersion: '2023-10-16' });
  }

  async createPaymentIntent(amount: number, currency: string = 'usd'): Promise<Stripe.PaymentIntent> {
    return this.stripe.paymentIntents.create({
      amount,
      currency,
    });
  }

  async createCustomer(email: string, name?: string): Promise<Stripe.Customer> {
    return this.stripe.customers.create({
      email,
      name,
    });
  }

  async createSubscription(
    customerId: string,
    priceId: string
  ): Promise<Stripe.Subscription> {
    return this.stripe.subscriptions.create({
      customer: customerId,
      items: [{ price: priceId }],
    });
  }

  async cancelSubscription(subscriptionId: string): Promise<Stripe.Subscription> {
    return this.stripe.subscriptions.cancel(subscriptionId);
  }

  async createCheckoutSession(
    priceId: string,
    successUrl: string,
    cancelUrl: string
  ): Promise<Stripe.Checkout.Session> {
    return this.stripe.checkout.sessions.create({
      mode: 'payment',
      line_items: [{ price: priceId, quantity: 1 }],
      success_url: successUrl,
      cancel_url: cancelUrl,
    });
  }

  async refundPayment(paymentIntentId: string): Promise<Stripe.Refund> {
    return this.stripe.refunds.create({
      payment_intent: paymentIntentId,
    });
  }

  async getPaymentIntent(paymentIntentId: string): Promise<Stripe.PaymentIntent> {
    return this.stripe.paymentIntents.retrieve(paymentIntentId);
  }

  async listCustomerSubscriptions(customerId: string): Promise<Stripe.ApiList<Stripe.Subscription>> {
    return this.stripe.subscriptions.list({ customer: customerId });
  }
}

File: services/payment/stripe_service.rb
require 'stripe'

module PaymentService
  class StripeService
    def initialize
      Stripe.api_key = ENV['STRIPE_SECRET_KEY']
    end

    def create_customer(email, name)
      Stripe::Customer.create(
        email: email,
        name: name
      )
    end

    def create_payment_intent(amount, currency, customer_id)
      Stripe::PaymentIntent.create(
        amount: amount,
        currency: currency,
        customer: customer_id,
        payment_method_types: ['card']
      )
    end

    def create_subscription(customer_id, price_id)
      Stripe::Subscription.create(
        customer: customer_id,
        items: [{ price: price_id }]
      )
    end

    def cancel_subscription(subscription_id)
      subscription = Stripe::Subscription.retrieve(subscription_id)
      subscription.cancel
    end

    def refund_payment(payment_intent_id)
      Stripe::Refund.create(payment_intent: payment_intent_id)
    end
  end
end
